### PR TITLE
GOBBLIN-1162: Provide an option to allow slow containers to commit su…

### PR DESCRIPTION
**Pull request by sv2000**
…icide when unhealthy

Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1162

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
In execution environments such as Gobblin-on-Yarn, where Gobblin workers can be re-assigned when the worker dies or is killed, it is useful to add a mode where each Gobblin task running inside a Gobblin worker can perform application-level health checks and report results of the health checks back to the worker hosting the tasks. The worker on receiving a health check failure event can be configured to exit the JVM. In the case of Gobblin-on-Yarn mode, this will result in the worker getting re-assigned to a different node. The proposed behavior is similar in flavor to speculative execution modes provided in other execution frameworks such as MapReduce. 

This change also provides an example of such an application-level health check that arises in the case of Kafka ingestion. In addition, this PR adds an EventBusFactory as an implementation of SharedResourcesBrokerFactory that allows communication between different components of GobblinTaskRunner via a shared EventBus instance.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Added unit tests:
1.EventBusFactoryTest
2. GobblinTaskRunnerTest
3. KafkaIngestionHealthCheckTest

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-api/src/main/java/org/apache/gobblin/commit/NoopCommitStep.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ */
+package org.apache.gobblin.commit;
+
+import java.io.IOException;
+
+
+/**
+ * A no-op implementation of {@link CommitStep}.
+ */
+public class NoopCommitStep implements CommitStep {
```
[the remark this answers is not in the document]
Comment (sv2000) [reply gobblin-api/src/main/java/org/apache/gobblin/commit/NoopCommitStep.java]: I had added it for testing. Removed it. Thanks!

File: gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/source/extractor/extract/kafka/KafkaIngestionHealthCheck.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Determine whether the commit step has been completed.
+   */
+  @Override
+  public boolean isCompleted()
```
[the remark this answers is not in the document]
Comment (sv2000) [reply gobblin-modules/gobblin-kafka-common/src/main/java/org/apache/gobblin/source/extractor/extract/kafka/KafkaIngestionHealthCheck.java:111]: Set it to false both here and inside HelixParticipantCheck. In general, the isCompleted() check is not useful for streaming use cases, where a CommitStep is going to be called multiple times. Further, even for non-streaming use cases, the isCompeted() check is useful only if there is a side-effect due to a partial execution of a CommitStep (e.g. renaming a subset of files from a file list).

File: gobblin-yarn/src/main/java/org/apache/gobblin/yarn/GobblinYarnTaskRunner.java [hunk 3]
```diff
@@ -209,6 +210,12 @@ public static void main(String[] args) throws Exception {
     } catch (ParseException pe) {
       printUsage(options);
       System.exit(1);
+    } catch (ContainerHealthCheckException e) {
+      // Ideally, we should not be catching this exception, as this is indicative of a non-recoverable exception. However,
+      // simply propagating the exception may prevent the container exit due to the presence of non-daemon threads present
+      // in the application. Hence, we catch this exception to invoke System.exit() which in turn ensures that all non-daemon threads are killed.
+      LOGGER.error("Exception encountered: {}", e);
+      System.exit(1);
```
[the remark this answers is not in the document]
Comment (sv2000) [reply gobblin-yarn/src/main/java/org/apache/gobblin/yarn/GobblinYarnTaskRunner.java:218]: The System.exit(1) will cause the JVM container running the GobblinYarnTaskRunner to terminate. The AMRM client running inside YarnService will then get a callback with ContainerExitStatus == ABORTED. For this case, the hostAffinity check will return to false, implying that the replacement container can be allocated on a different physical host.

[the remark this answers is not in the document]
Comment (autumnust) [reply gobblin-yarn/src/main/java/org/apache/gobblin/yarn/GobblinYarnTaskRunner.java:218]: Sounds good !

**Pull request closed without merging**